lib/render/mayaToHydra/renderOverride.cpp

```diff
@@ -0,0 +1,842 @@
+        : UFE_NS::Observer(), _mtohRenderOverride(mtohRenderOverride) {}
+
+    //~UfeSelectionObserver() override {}
+
+    void operator()(const UFE_NS::Notification& notification) override {
+        // During Maya file read, each node will be selected in turn, so we get
+        // notified for each node in the scene.  Prune this out.
+        if (MFileIO::isOpeningFile()) { return; }
+
+        auto selectionChanged =
+            dynamic_cast<const UFE_NS::SelectionChanged*>(&notification);
+        if (selectionChanged == nullptr) { return; }
+
+        TF_DEBUG(HDMAYA_RENDEROVERRIDE_SELECTION)
+            .Msg(
+                "UfeSelectionObserver triggered (ufe selection change "
+                "triggered)\n");
+        _mtohRenderOverride.SelectionChanged();
+    }
+
+private:
+    MtohRenderOverride& _mtohRenderOverride;
+};
+
+#endif // WANT_UFE_BUILD
+
+/// Simple RAII class to save uniform buffer bindings, to deal with a maya
+/// issue.
+///
+/// As originally explained by Pixar in their usdmaya plugin:
+///
+/// XXX: When Maya is using OpenGL Core Profile as the rendering engine (in
+/// either compatibility or strict mode), batch renders like those done in
+/// the "Render View" window or through the ogsRender command do not
+/// properly track uniform buffer binding state. This was causing issues
+/// where the first batch render performed would look correct, but then all
+/// subsequent renders done in that Maya session would be completely black
+/// (no alpha), even if the frame contained only Maya-native geometry or if
+/// a new scene was created/opened.
+///
+/// To avoid this problem, we need to save and restore Maya's bindings
+/// across Hydra calls. We try not to bog down performance by saving and
+/// restoring *all* GL_MAX_UNIFORM_BUFFER_BINDINGS possible bindings, so
+/// instead we only do just enough to avoid issues. Empirically, the
+/// problematic binding has been the material binding at index 4.
+class UBOBindingsSaver {
+public:
+    static constexpr size_t UNIFORM_BINDINGS_TO_SAVE = 5u;
+
+    UBOBindingsSaver() {
+        for (size_t i = 0u; i < _uniformBufferBindings.size(); ++i) {
+            glGetIntegeri_v(
+                GL_UNIFORM_BUFFER_BINDING, (GLuint)i,
+                &_uniformBufferBindings[i]);
+        }
+    }
+
+    ~UBOBindingsSaver() {
+        for (size_t i = 0u; i < _uniformBufferBindings.size(); ++i) {
+            glBindBufferBase(
+                GL_UNIFORM_BUFFER, (GLuint)i, _uniformBufferBindings[i]);
+        }
+    }
+
+private:
+    std::array<GLint, UNIFORM_BINDINGS_TO_SAVE> _uniformBufferBindings;
+};
+
+} // namespace
+
+MtohRenderOverride::MtohRenderOverride(const MtohRendererDescription& desc)
+    : MHWRender::MRenderOverride(desc.overrideName.GetText()),
+      _rendererDesc(desc),
+      _selectionTracker(new HdxSelectionTracker),
+      _renderCollection(
+          HdTokens->geometry,
+          HdReprSelector(
+#if MAYA_APP_VERSION >= 2019
```

Comment:
Why is there conditional compilation on maya version to determine different Hydra code path?

Reply:
So Pal made a commit deliberately disabling smooth-mesh preview support for < 2019:

https://github.com/Autodesk/maya-usd/commit/6f981278479d6b4566580ea892d2f806786808a1

Unfortunately, when I asked him, he wasn't quite sure why anymore.  He thinks he was getting some sort of crash...?

Reply:
Okay, then we will have to keep it the way it is.